core/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -152,7 +152,7 @@ public void search(Query query, Collector collector) throws IOException {
             // throws Lucene.EarlyTerminationException when given count is reached
             collector = Lucene.wrapCountBasedEarlyTerminatingCollector(collector, searchContext.terminateAfter());
         }
-        if (currentState == Stage.MAIN_QUERY) {
+        if (currentState == Stage.MAIN_QUERY && (query == searchContext.query() || query == searchContext.parsedQuery().query())) {
```

Comment:
this part of the change looks a bit scary to me. I'm wondering how hard it would be to do this 'state' management and wrapping of collectors outside of IndexSearcher so that we don't have to check what the query is.